src/grpc/router_ics_devaddr_worker.erl

```diff
@@ -0,0 +1,262 @@
+    {reply, ok, State}.
+
+handle_cast({?RECONCILE_START, Pid}, #state{conn_backoff = Backoff0} = State) ->
+    case get_devaddrs(Pid, State) of
+        {error, _Reason} = Error ->
+            {Delay, Backoff1} = backoff:fail(Backoff0),
+            _ = erlang:send_after(Delay, self(), ?INIT),
+            lager:warning("fail to get_devaddrs ~p, retrying in ~wms", [
+                _Reason, Delay
+            ]),
+            ok = forward_reconcile(Pid, Error),
+            {noreply, State#state{conn_backoff = Backoff1}};
+        {ok, _Stream} ->
+            {_, Backoff2} = backoff:succeed(Backoff0),
+            {noreply, State#state{conn_backoff = Backoff2}}
+    end;
+handle_cast({?RECONCILE_END, Pid, DevaddrRanges}, #state{} = State) ->
+    ok = forward_reconcile(Pid, DevaddrRanges),
+    %% Drop ranges that may fall outside the configured devaddr_prefix
+    Ranges = lists:filtermap(
+        fun(DevaddrRange) ->
+            #iot_config_devaddr_range_v1_pb{
+                start_addr = StartAddr,
+                end_addr = EndAddr
+            } = DevaddrRange,
+            try
+                MinBase = router_ics_devaddr_worker:devaddr_num_to_base_num(StartAddr),
+                MaxBase = router_ics_devaddr_worker:devaddr_num_to_base_num(EndAddr),
+                {true, {MinBase, MaxBase}}
+            catch
+                _Error:Reason ->
+                    lager:warning("ignoring devaddr range [reason: ~p]", [Reason]),
+                    false
+            end
+        end,
+        DevaddrRanges
+    ),
+    ok = router_device_devaddr:set_devaddr_bases(Ranges),
+
+    {noreply, State#state{devaddr_ranges = DevaddrRanges}};
+handle_cast(_Msg, State) ->
+    lager:warning("rcvd unknown cast msg: ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info(?INIT, #state{conn_backoff = Backoff0, route_id = RouteID} = State) ->
+    {Delay, Backoff1} = backoff:fail(Backoff0),
+    case connect(State) of
+        {error, _Reason} ->
+            lager:warning("fail to connect ~p, reconnecting in ~wms", [_Reason, Delay]),
+            _ = erlang:send_after(Delay, self(), ?INIT),
+            {noreply, State#state{conn_backoff = Backoff1}};
+        ok ->
+            lager:info("connected"),
+            {_, Backoff2} = backoff:succeed(Backoff0),
+            ok = ?MODULE:reconcile(undefined),
+            {noreply, State#state{
+                conn_backoff = Backoff2, route_id = RouteID
+            }}
+    end;
+handle_info({headers, _StreamID, _Data}, State) ->
+    ct:print("got headers for stream: ~p, ~p", [_StreamID, _Data]),
+    {noreply, State};
+handle_info({trailers, _StreamID, _Data}, State) ->
+    ct:print("got trailers for stream: ~p, ~p", [_StreamID, _Data]),
+    {noreply, State};
+handle_info({eos, _StreamID}, State) ->
+    ct:print("got eos for stream: ~p", [_StreamID]),
+    {noreply, State};
+handle_info({'DOWN', _Ref, Type, Pid, Reason}, State) ->
```

Comment:
you can also add end stream here see https://github.com/helium/router/commit/3ddf73c837c6f4d4d5118de8eae601f7aab16e6d#diff-3ef05d1bad48450f377828f9de8884492efe107bda131f41b4319a0c8f3d480cR240